File: leetcode/LC/graph/dijkstra/0743_network_delay_time.py
from typing import List
from collections import heapq

class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        """
        1. create adjacency matrix
        2. initialize min heap
        3. create distance to store final weight at that index node, initialize it with infinity
        """
        # create adjacency list with initial node as graph index and values as vector of pairs with values target node and weight.
        adj_list = [{} for _ in range(n+1)]
        
        for u,v, w in times:
            adj_list[u][v] = w
        
        minheap = [(0, k)]
        distance = [float('inf') for _ in range(n+1)]
        
        # 0 is taken extra for indexing purpose
        distance[0] = 0
        
        # distance of source is 0
        distance[k] = 0
        
        while minheap:
            wt, n = heapq.heappop(minheap)
            for target, weight in adj_list[n].items():
                if distance[target] > wt + weight:
                    heapq.heappush(minheap, (wt+weight, target))
                    distance[target] = wt + weight
        if max(distance) == float('inf'): return -1
        return max(distance)
        

File: leetcode/LC1/graph/0743_network_delay_time.cpp
#include <climits>
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <set>
#include <unordered_map>
using namespace std;

//dijkastra algorithm, also check heap approach https://leetcode.com/problems/network-delay-time/solution/

class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        // create adjacency list with initial node as graph index and values as vector of pairs with values target node and weight.
        vector<vector<pair<int, int>>>graph(n+1); // {targetnode, wt}
        for(auto time: times){
            graph[time[0]].push_back({time[1], time[2]});
        }
        
        // set distance of all node from target nodes to infinity
        vector<int> distance(n+1, INT_MAX);
        // take ordered set to get node with minimum weight till now.
        set <pair<int,int>> s; // {wt, target node}
        s.insert({0, k});
        distance[k] = 0;
        distance[0] = 0;
        while(!s.empty()){
            auto x = *(s.begin());
            s.erase(x);
            for(auto time: graph[x.second]){
                //cout<< time.first <<" "<<time.second<<endl;
                if(distance[time.first] > x.first + time.second){
                    distance[time.first] = x.first + time.second;
                    s.insert({distance[time.first], time.first});
                }
            }
        }
        int ans = *max_element(distance.begin(), distance.end());
        if (ans == INT_MAX)ans = -1;
        return ans;
    }
};
